emacs/tests/merlin-cap-test.el

```diff
@@ -0,0 +1,252 @@
+
+;;; Code:
+
+(require 'merlin)
+(require 'ert)
+
+(ert-deftest test-merlin-cap--bounds ()
+  (should (equal (merlin-cap--regions "Aaa.bbb.c" "cc.ddd")
+                 '("Aaa.bbb." "Aaa.bbb." "ccc." "ddd")))
+  (should (equal (merlin-cap--regions "~fo" "o.bar")
+                 '("" "" "~foo" "")))
+  (should (equal (merlin-cap--regions "" "~foo.bar")
+                 '("" "" "~foo" "")))
+  (should (equal (merlin-cap--regions "~fo" "o~bar")
+                 '("" "" "~foo" "")))
+  (should (equal (merlin-cap--regions "~foo" "~bar")
+                 '("" "" "~foo" "")))
+  (should (equal (merlin-cap--regions "~fo" "o.b~ar")
+                 '("" "" "~foo" "")))
+  ;; There's no obvious correct thing to return in this case, so this is fine.
+  (should (equal (merlin-cap--regions "~foo.bar" "")
+                 '("foo." "foo." "bar" "")))
+  (should (equal (merlin-cap--regions "" "~")
+                 '("" "" "~" "")))
+  (should (equal (merlin-cap--regions "" "Aaa.bbb.ccc.ddd")
+                 '("" "" "Aaa." "bbb.ccc.ddd")))
+  (should (equal (merlin-cap--regions "A" "aa.bbb.ccc.ddd")
+                 '("" "" "Aaa." "bbb.ccc.ddd")))
+  ;; An "atom" can also just be a dotted path projecting from an expression
+  (should (equal (merlin-cap--regions "(foo bar)." "")
+                 '("." "." "" "")))
+  (should (equal (merlin-cap--regions "(foo bar).Aa" "a")
+                 '("." "." "Aaa" "")))
+  (should (equal (merlin-cap--regions "(foo bar).Aaa.Bb" "b.ccc")
+                 '("." ".Aaa." "Bbb." "ccc")))
+  (should (equal (merlin-cap--regions "(foo bar).Aaa.bb" "b.ccc")
+                 '("." ".Aaa." "bbb." "ccc")))
+  (should (equal (merlin-cap--regions "(foo bar).aaa.bb" "b.ccc")
+                 '(".aaa." ".aaa." "bbb." "ccc")))
+  ;; We should omit only uppercase components before point, not lowercase ones
+  (should (equal (merlin-cap--regions "M." "x")
+                 '("" "M." "x" "")))
+  (should (equal (merlin-cap--regions "M.t." "x")
+                 '("M.t." "M.t." "x" "")))
+  (should (equal (merlin-cap--regions "M.N." "x")
+                 '("" "M.N." "x" "")))
+  (should (equal (merlin-cap--regions "M.t.N." "x")
+                 '("M.t." "M.t.N." "x" "")))
+  (should (equal (merlin-cap--regions "aa.bB.CC.x" "")
+                 '("aa.bB." "aa.bB.CC." "x" "")))
+  (should (equal (merlin-cap--regions "Aa.bB.CC.x" "")
+                 '("Aa.bB." "Aa.bB.CC." "x" "")))
+  (should (equal (merlin-cap--regions "aa.Bb.cc.x" "")
+                 '("aa.Bb.cc." "aa.Bb.cc." "x" "")))
+  (should (equal (merlin-cap--regions "aa.Bb.Cc.x" "")
+                 '("aa." "aa.Bb.Cc." "x" ""))))
+
+(defvar-local messages-buffer-name "*Messages*")
+
+(defun merlin-cap--current-message ()
+  "Like `current-message' but work in batch mode and use `messages-buffer-name'."
+  (with-current-buffer messages-buffer-name
+    (save-excursion
+      (forward-line -1)
+      (buffer-substring (point) (line-end-position)))))
```

Comment:
I changed `pos-eol` to `line-end-position` this might not be necessary with `compat` package? 
```suggestion
      (buffer-substring (point) (pos-eol)))))
```